### PR TITLE
add RFC for script run hooks

### PR DESCRIPTION
**Pull request by deecewan**
This was spurred by https://github.com/yarnpkg/yarn/issues/3209#issuecomment-308606581 to try and normalise the hooks that run and when.

### CODE REVIEW

File: text/0000-update-hook-runs.md [hunk 1]
```diff
@@ -0,0 +1,196 @@
...
+  - runs *before* the version commit is made
+- `postversion`
+  - after `version`
+  - runs *after* the version commit is made
+- `preshrinkwrap`
+  - before `shrinkwrap`
+- `shrinkwrap`
+  - before `shrinkwrap`
+    - confusing, but runs *before* the shrinkwrap is created
+  - I'm not sure `yarn` needs to include this, given that `shrinkwrap` is not
```
Comment (bestander) [inline text/0000-update-hook-runs.md]: yeah, we don't need shrinkwraps.

**Comment by bestander** [comment]
Great work, @deecewan!

**Comment by bestander** [comment]
Let me summon a few folks who have more context

**Comment by bestander** [comment]
My opinion - this looks fine, we should just match npm@5 and move forward.
Are there any tricks we need to address here?
I remember seeing a few issues where people are expecting to run post/preinstall scripts when reinstalling node_modules.

**Comment by deecewan** [comment]
AFAIK, from the docs and from testing, I've enumerated how everything works. I might make a couple of test scripts for `npm@5` to 'codify' the expected behaviours.

**Comment by bestander** [comment]
Great job, @deecewan!
Let's do it! Are you up to send a few PRs?
Tests would be very important to make sure we don't regress

**Comment by deecewan** [comment]
yeah, keen!

I'll do what I can 👍

**Comment by andreineculau** [comment]
FYI a programmatic comparison of which hooks run and in what order, along with the toold can now be found at https://github.com/andreineculau/package-json-scripts

**Comment by bestander** [comment]
@deecewan, any chance you've had a crack at this?
@andreineculau, do you want to help out?

File: text/0000-update-hook-runs.md (hunk 1 above)
Comment (andreineculau) [inline text/0000-update-hook-runs.md]: the shrinkwrap scripts also run when creating the new `package-lock.json`, so yarn should also run if when creating the `yarn.lock` file, for a closer 1-to-1 match

**Comment by andreineculau** [comment]
@bestander I'm both limited by time and incentive as I do not use yarn in my routine. I've done the package-json-scripts investigation because we intend to have git dependencies, and I've seen inconsistencies. My first npm alternative to try out is pnpm actually, so trying out and even committing to yarn is too much in the future.

What I can help out with, no doubt, is doing one final manual testing of whatever new behaviour lands or is about to land in yarn, if nobody wants to run https://github.com/andreineculau/package-json-scripts themselves.

**Comment by deecewan** [comment]
@bestander i started taking a look at this. `yarn` already does 95% of theses hooks in the correct order. I went through and figured out which hooks run and when, and then went through yarn and checked on them.

from memory, i think there were only 1 or 2 changes. i'll take another look after work tonight.

**Comment by bestander** [comment]
@deecewan, awesome, thank you very much.
Do you think you could set up a couple of tests that cover this contract?

**Pull request merged**